apps/frontend/src/lib/psi/index.ts

```diff
@@ -22,20 +23,81 @@ export const generatePSIKeyPair = async () => {
   };
 };
 
-export const generateBitVectorFromLannaData = (
-  data: LannaData
-): Uint32Array => {
-  const bitVector = new Uint32Array(1000).fill(0);
-
-  // 0-7 reserved for desired connections
-  bitVector[0] = data.desiredConnections.getHealthy ? 1 : 0;
-  bitVector[1] = data.desiredConnections.cowork ? 1 : 0;
-  bitVector[2] = data.desiredConnections.enjoyMeals ? 1 : 0;
-  bitVector[3] = data.desiredConnections.learnFrontierTopics ? 1 : 0;
-  bitVector[4] = data.desiredConnections.findCollaborators ? 1 : 0;
-  bitVector[5] = data.desiredConnections.goExploring ? 1 : 0;
-  bitVector[6] = data.desiredConnections.party ? 1 : 0;
-  bitVector[7] = data.desiredConnections.doMentalWorkouts ? 1 : 0;
-
-  return bitVector;
+const PSI_BIT_VECTOR_SIZE = 10000;
+const COMMON_CONNECTIONS_BIT_VECTOR_SIZE = 9000;
+
+export const generateBitVectorFromUserData = (
+  userData: UserData,
+  connections: Record<string, Connection>
+): { bitVector: Uint32Array; indexMapping: Record<number, string[]> } => {
+  const bitVector = new Uint32Array(PSI_BIT_VECTOR_SIZE).fill(0);
+  const bitIndexToUsername: Record<number, string[]> = {};
+
+  // 0-8999 reserved for common connections
+  for (const connection of Object.values(connections)) {
+    const { username, signaturePublicKey } = connection.user;
+    const hash = sha256(signaturePublicKey);
+    const bigIntHash = BigInt("0x" + hash);
+    const index = Number(
+      bigIntHash % BigInt(COMMON_CONNECTIONS_BIT_VECTOR_SIZE)
+    );
+    bitVector[index] = 1;
```

Comment:
What prevents multiple connections from not overlapping on the same index? The large denominator (9000)?

Reply:
When a user is created could they be allocated an incremented index that's used for their bitvector position? 

The middleman can already calculate the index in this scheme (because the sigPubKey is known to them, right? Or is it fully encrypted?).

Reply:
yeah there's nothing here to prevent overlap. this bucketing method will lead to errors, which is just unlikely for now. but ux of increasing vector size does become worse